File: 322-coin-change/322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        
        dp = [float('inf') for _ in range(amount+1)]
        dp[0] = 0
        for coin in coins:
            for i in range(1, amount+1):
                if i >= coin:
                    dp[i] = min(dp[i], dp[i - coin] + 1)
        return dp[amount] if dp[amount] != float('inf') else -1

File: 322-coin-change/322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) 
    {
        vector <int> dp(amount+1, INT_MAX / 2);
        dp[0] = 0;
        for(auto coin : coins)
        {
            for(int i = 1; i < amount + 1; i++)
            {
                if(i >= coin)
                {
                    dp[i] = min(dp[i], 1 + dp[i-coin]);
                }
            }
        }
        
        return dp[amount] == INT_MAX/2 ? -1 : dp[amount];
    }
};